src/server/ua_server.c

```diff
@@ -620,6 +620,108 @@ UA_Server_createServerConnection(UA_Server *server, const UA_String *serverUrl)
     return UA_STATUSCODE_BADINTERNALERROR;
 }
 
+UA_StatusCode UA_Server_attemptReverseConnect(UA_Server *server, reverse_connect_context *context) {
+    UA_ServerConfig *config = UA_Server_getConfig(server);
+
+    UA_StatusCode res = UA_STATUSCODE_BADINTERNALERROR;
+    UA_String tcpString = UA_STRING_STATIC("tcp");
+    for(UA_EventSource *es = config->eventLoop->eventSources; es != NULL; es = es->next) {
+        /* Is this a usable connection manager? */
+        if(es->eventSourceType != UA_EVENTSOURCETYPE_CONNECTIONMANAGER)
+            continue;
+        UA_ConnectionManager *cm = (UA_ConnectionManager*)es;
+        if(!UA_String_equal(&tcpString, &cm->protocol))
+            continue;
+
+        if (es->state != UA_EVENTSOURCESTATE_STARTED)
+            return UA_STATUSCODE_GOODCOMPLETESASYNCHRONOUSLY;
+
+        /* Set up the parameters */
+        UA_KeyValuePair params[2];
+        size_t paramsSize = 2;
+
+        params[0].key = UA_QUALIFIEDNAME(0, "hostname");
+        UA_Variant_setScalar(&params[0].value, &context->hostname, &UA_TYPES[UA_TYPES_STRING]);
+
+        params[1].key = UA_QUALIFIEDNAME(0, "port");
+        UA_Variant_setScalar(&params[1].value, &context->port, &UA_TYPES[UA_TYPES_UINT16]);
+
+        /* Open the server connection */
+        res = cm->openConnection(cm, paramsSize, params, server, context,
+                                 UA_Server_reverseConnectCallback);
+
+        if (res != UA_STATUSCODE_GOOD) {
+            context->currentConnection = NULL;
+            context->status = REVERSE_CONNECT_STATUS_DISCONNECTED;
+        } else {
+            context->status = REVERSE_CONNECT_STATUS_CONNECTING;
+        }
+    }
+
+    return res;
+}
+
+UA_StatusCode UA_Server_addReverseConnect(UA_Server *server, UA_String url, UA_UInt64 *handle) {
```

Comment:
I miss the cleanup of pending reverse connects during server shutdown…
Is this via the closing tcp socket callback?

Reply:
Yes, this is done automatically when the socket state changes on shutdown.